src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -192,458 +96,48 @@ void IUtilities::updateImageViaStagingBuffer(
-                {
-                    assert(canTransferMipLevelsPartially);
-                    bool success = tryFillLayer();
-                    if(success)
-                        anyTransferRecorded = true;
-                    else
-                        break; // not enough mem -> break
-                }
-                // There is remaining rows in slice that needs copying
-                else if (currentBlockInRow == 0 && currentRowInSlice > 0)
-                {
-                    assert(canTransferMipLevelsPartially);
-                    bool success = tryFillSlice();
-                    if(success)
-                        anyTransferRecorded = true;
-                    else
-                        break; // not enough mem -> break
-                }
-                // There is remaining blocks in row that needs copying
-                else if (currentBlockInRow > 0)
+                asset::IImage::SBufferCopy nextRegionToCopy = {};
+                if(availableUploadBufferMemory > 0u && regionIterator.advanceAndCopyToStagingBuffer(nextRegionToCopy, availableUploadBufferMemory, currentUploadBufferOffset, m_defaultUploadBuffer->getBufferPointer()))
                 {
-                    assert(canTransferMipLevelsPartially);
-                    bool success = tryFillRow();
-                    if(success)
-                        anyTransferRecorded = true;
-                    else
-                        break; // not enough mem -> break
+                    anyTransferRecorded = true;
+                    regionsToCopy.push_back(nextRegionToCopy);
                 }
+                else
+                    break;
             }
 
             if(!regionsToCopy.empty())
             {
                 cmdbuf->copyBufferToImage(m_defaultUploadBuffer.get()->getBuffer(), dstImage, dstImageLayout, regionsToCopy.size(), regionsToCopy.data());
             }
 
-            assert(anyTransferRecorded && "uploadBufferSize is not enough to support the smallest possible transferable units to image, may be caused if your queueFam's minImageTransferGranularity is large or equal to <0,0,0>.");
+            assert(anyTransferRecorded && "allocationSize is not enough to support the smallest possible transferable units to image, may be caused if your queueFam's minImageTransferGranularity is large or equal to <0,0,0>.");
             
             // some platforms expose non-coherent host-visible GPU memory, so writes need to be flushed explicitly
             if (m_defaultUploadBuffer.get()->needsManualFlushOrInvalidate()) {
-                IDeviceMemoryAllocation::MappedMemoryRange flushRange(m_defaultUploadBuffer.get()->getBuffer()->getBoundMemory(), localOffset, uploadBufferSize);
+                IDeviceMemoryAllocation::MappedMemoryRange flushRange(m_defaultUploadBuffer.get()->getBuffer()->getBoundMemory(), localOffset, allocationSize);
```

Comment:
you only need to flush what you've written (rounded up to coherent atom size), not the whole allocation

Reply:
Good point, another place where I just used the code as before and didn't think twice